### ISSUE TITLE
Issues with German spelling

### ISSUE THREAD

**Issue by laino** [vim/vim#3587]
There appears to be something wrong with ``de_20`` and ``de``:

At the very least it does not correctly detect *desweiteren* as incorrect.

I read somewhere that spell files are imported from OpenOffice, but a quick
test in LibreOffice revealed that spell checking is working fine there.

It's kinda annoying since that word is an extremely common misspelling.

Other German spell files are unaffected.

**Comment by brammool** [comment]
> There appears to be something wrong with ``de_20`` and ``de``:
> 
> At the very least it does not correctly detect *desweiteren* as incorrect.
> 
> I read somewhere that spell files are imported from OpenOffice, but a quick
> test in LibreOffice revealed that spell checking is working fine there.
> 
> Other German spell files are unaffected. 

The German spell files were last updated in 2010.  Someone should spend
a bit of time on getting the latest version and applying the changes to
make them work in Vim.

-- 
   An extraordinary TALL KNIGHT in all black (possibly John with Mike on his
   shoulders) walks out from the dark trees.  He is extremely fierce and
   gruesome countenance.  He walks towards KING ARTHUR and PATSY, who are
   wazzing like mad.  (Salopian slang, meaning very scared.  almost to the
   point of wetting oneself, e.g. before an important football match or
   prior to a postering.  Salopian slang meaning a beating by the school
   praeposters.  Sorry about the Salopian slant to this stage direction - Ed.)
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
A `:spelldump` of de_20 spelllang shows:

```
/regions=de1920atch
.....
deswegen
desweiteren/13
desweiterend/13
```

According to `:h spell-wordlist-format`, the `/13` specifies a region where this is valid. According to the header, regions 1 and 3 would mean language `de` and `de_20`. However, that is certainly wrong, as the [Duden states](https://www.duden.de/suchen/dudenonline/desweiteren).

**Comment by chrisbra** [comment]
> The German spell files were last updated in 2010. Someone should spend a bit of time on getting the latest version and applying the changes to make them work in Vim.

It was always wrong however :)

Where is documented, what needs to be done to get updated spellfiles? I can do it.

**Comment by brammool** [comment]
Christian wrote:

> > The German spell files were last updated in 2010. Someone should
> > spend a bit of time on getting the latest version and applying the
> > changes to make them work in Vim.
> 
> It was always wrong however :)
> 
> Where is documented, what needs to be done to get updated spellfiles?
> I can do it.

runtime/spell/README.txt

You will have to figure out the right place to download from, it has
probably changed.  The old ones can still be found at
http://download.services.openoffice.org/files/contrib/dictionaries/
They date back to 2006, there must be newer ones.

Then make sure the differences in runtime/spell/de/*.diff are applied,
you probably will have a make manual changes.

It's not required to use Aap for doing this.  You could also drop the
main.aap file and use manual instructions, since it's done infrequently.
But if you do manage to make the main.aap work it's easier for me to
create the distribution.

-- 
TALL KNIGHT: We shall say Ni! again to you if you do not appease us.
ARTHUR:      All right!  What do you want?
TALL KNIGHT: We want ... a shrubbery!
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
Okay, thanks for the pointers. How about those patches:
https://github.com/chrisbra/vim/compare/5608cce08057cd243435259a9fd3e867cd9d7a8f...6270028d2b75bb63e8f9318798bc9f53c5d14da9

(I also committed the resulting spl and sug files, so you should be able to just download them).

There doesn't seem to exist a combined dictionary for de_de anymore. I therefore used the new orthography for de_DE and I would also propose that spellang=de uses only the new spelling and not the old anymore. Not sure, if this needs any adjustments on Vims side however. I am not entirely sure, how from the different regions we come to the spelllang settings `de de_de de_19 de_20`.

**Comment by tonymec** [comment]
Experiments with a few words known to differ in _19 and _20 ought to show which is which. What about de_AT vocabulary (with e.g. Erdapfel instead of Kartoffel) and de_CH orthography (with ss replacing ß everywhere, as in "Wie trinken die Schweizer Bier? — In Masse"): Are they included in plain "de"?

**Comment by brammool** [comment]
Christian wrote:

> Okay, thanks for the pointers. How about those patches:
> https://github.com/chrisbra/vim/compare/5608cce08057cd243435259a9fd3e867cd9d7a8f...6270028d2b75bb63e8f9318798bc9f53c5d14da9
> 
> (I also committed the resulting spl and sug files, so you should be able to just download them).

Thanks.  Looking at the diff, the resulting .diff files are huge.  That
can't be right, they should mainly contain changes to the .aff header,
adding the SAL lines and such.  Perhaps the .aff and .aff.orig files are
from different versions?  The idea is to delete the .aff.orig files,
download the new spell files, then copy .aff to .aff.orig and then make
the changes needed for Vim.  After that the .aff.diff can be created, so
that it can be applied by someone else.

How do I see the resulting files?  Github doesn't appear to have a link
to that.


> There doesn't seem to exist a combined dictionary for de_de anymore. I
> therefore used the new orthography for de_DE and I would also propose
> that spellang=de uses only the new spelling and not the old anymore.
> Not sure, if this needs any adjustments on Vims side however. I am not
> entirely sure, how from the different regions we come to the spelllang
> settings `de de_de de_19 de_20`.

Well, what would most users want to use?  That is what "de" should
result in.

-- 
You were lucky. We lived for three months in a brown paper bag in a 
septic tank. We used to have to get up at six o'clock in the morning, 
clean the bag, eat a crust of stale bread, go to work down mill for 
fourteen hours a day week in-week out. When we got home, our Dad
would thrash us to sleep with his belt!

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
> Thanks. Looking at the diff, the resulting .diff files are huge.

I deleted duplicates (or whatever Vim thought of duplicates from the dic file. That were many. The change to the aff file was minimal, mainly dropping lines that Vim does not understand. There were a couple of conversion warning when run in latin1 environment, but that should not harm.

> How do I see the resulting files? Github doesn't appear to have a link to that.

If you click on the file, (e.g. https://github.com/chrisbra/vim/blob/master/runtime/spell/de.utf-8.spl, you can click to download it).

> Well, what would most users want to use? That is what "de" should result in.

Considering that the reformation of the spelling has been done more than twenty years ago and has been in effect for more than 10 years, I would go so far and say it should only be the new german spelling and the old one should not be used anymore.

**Comment by brammool** [comment]
Christian wrote:

> > Thanks. Looking at the diff, the resulting .diff files are huge.
> 
> I deleted duplicates (or whatever Vim thought of duplicates from the
> dic file. That were many. The change to the aff file was minimal,
> mainly dropping lines that Vim does not understand. There were a
> couple of conversion warning when run in latin1 environment, but that
> should not harm.
> 
> > How do I see the resulting files? Github doesn't appear to have a link to that.
> 
> If you click on the file, (e.g.
> https://github.com/chrisbra/vim/blob/master/runtime/spell/de.utf-8.spl,
> you can click to download it).

I wanted to see the input files, these are apparently at
https://github.com/chrisbra/vim/blob/master/runtime/spell/de/

It appears the SAL lines are missing.  These are essential for
suggesting the right word.  Check the previous .diff files for what was
added specifically for Vim.

A lot of lines in the .aff file are deleted, like:

	 Abitur/EPSm
	-Abitur/Smij

I wonder why these duplicates even exist.  I would think it should be a
combination of all flags:

	 Abitur/EPSmij

But it's been a while since I looked at spell stuff.  If this what is
supposed to happen, instead of changing the .dic file, we could change
the code that reads the file to work around the duplicates with
different flags.  Otherwise we get huge .diff files.


> > Well, what would most users want to use? That is what "de" should result in.
> 
> Considering that the reformation of the spelling has been done more
> than twenty years ago and has been in effect for more than 10 years, I
> would go so far and say it should only be the new german spelling and
> the old one should not be used anymore.

OK, you know this better than me.

-- 
You got to work at a mill?  Lucky!  I got sent back to work in the
acid-mines for my daily crust of stale bread... which not even the
birds would eat.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
> It appears the SAL lines are missing. These are essential for suggesting the right word. Check the previous .diff files for what was added specifically for Vim.

Where did those lines originally come from? They are not part of the dictionary .aff files (https://cgit.freedesktop.org/libreoffice/dictionaries/tree/de/de_DE_frami.aff)

I can add those back, but would like to know where they came from initially.
I'll have a look at it later tomorrow.

> we could change the code that reads the file to work around the duplicates with different flags. Otherwise we get huge .diff files.

Or we can just ignore the Warning...

**Comment by brammool** [comment]
Christian wrote:

> > It appears the SAL lines are missing. These are essential for
> > suggesting the right word. Check the previous .diff files for what
> > was added specifically for Vim.
> 
> Where did those lines originally come from? They are not part of the dictionary .aff files (https://cgit.freedesktop.org/libreoffice/dictionaries/tree/de/de_DE_frami.aff)
> 
> I can add those back, but would like to know where they came from initially.
> I'll have a look at it later tomorrow.

The German SAL lines come from aspell, it's in the comment.  They
implement sound folding.  See :help spell-SAL

> > we could change the code that reads the file to work around the
> > duplicates with different flags. Otherwise we get huge .diff files.
> 
> Or we can just ignore the Warning...

Hmm, they are indeed duplicates, since the same flag is often repeated.
So the warning is appropriate in many lines.

-- 
Apologies for taking up the bandwidth with the apology.  Anything else I
can apologise for ...... er no can't think of anything, sorry about that.
		Andy Hunt (Member of British Olympic Apology Squad)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by chrisbra** [comment]
Okay, I changed it according to your comments:
https://github.com/chrisbra/vim/compare/5e021c990f8817a50d3264782a5...3b27c92f297540761ebbd92d04fb3

> Hmm, they are indeed duplicates, since the same flag is often repeated. So the warning is appropriate in many lines.

Yeah, but Vim only throws the first warning and then shuts up (however this is a bit annoying when you try to fix it, because you need to run `:mkspell` several times until no duplicates are there anymore). But I don't care currently.

**Comment by chrisbra** [comment]
fixed as of 2a953fcf107d24229fec8af820ee62c908caafbf
